### ISSUE TITLE
Type narrowing and reachability differ for `isinstance` with tuple argument

### ISSUE THREAD

**Issue by tingerrr** [astral-sh/ty#4021]
### Summary

It seems that there is a difference between reachability and type narrowing when `isinstance` is used with a tuple of type variables. Type narrowing seems correct in both cases, in both cases `other` is `Never` the divergent branch. However, reachability analysis reports an implicit `None`-return in the second case.

Note that if `reveal_type(other)` is not commented out, both cases type check fine. My guess is that reachability takes uses of `reveal_type` into account, which in turn get their results from narrowing?

```py
from dataclasses import dataclass

class Foo:
    def func(self, other: FooOrBar) -> bool:
        return True

class Bar:
    def func(self, other: FooOrBar) -> bool:
        return True

FooOrBar = Foo | Bar

def func(self, other: Foo) -> bool:
    # if isinstance(other, Foo) or isinstance(other, Bar):
    if isinstance(other, (Foo, Bar)):
        return self.value.foo(other)

    # reveal_type(other)
```

Playground link: https://play.ty.dev/87afff5e-0eb2-4121-a5ec-a3415520b3d7

### Version

ty (543ea43a5)

**Comment by sharkdp** [comment]
Thank you for reporting this.

As discussed on Discord, we need to infer `isinstance(other, (Foo, Bar))` as `Literal[True]` here in order for this to work. We already have special inference support for `isinstance`. We probably need to extend it to also handle tuple arguments.
